expression/builtin_cast.go

```diff
@@ -688,7 +691,8 @@ func (b *builtinCastStringAsJSONSig) evalJSON(row chunk.Row) (res json.BinaryJSO
 	if isNull || err != nil {
 		return res, isNull, err
 	}
-	if mysql.HasParseToJSONFlag(b.tp.Flag) {
+	data := hack.Slice(val)
+	if mysql.HasParseToJSONFlag(b.tp.Flag) && json2.Valid(data) {
```

Comment:
Probabily we don't need use `encoding/json` to check the validity, just ParseBInaryFromString and if err is not nil we will fall back to CreateBinary?